### DISCUSSION TITLE
What is your preferred way of using Copilot and cmp?

### DISCUSSION THREAD

**Discussion by ogirginc** [hrsh7th/nvim-cmp#1034] [Q&A]
Do you use Copilot as a `nvim-cmp` source? Or do you have an alternative usage where Copilot is treated something different with different key bindings, looks, etc.?

**Comment by belltoy** [comment]
Copilot is not like the traditional complete components. What the traditional tools provide is to suggest a word, may be based on current/previous input chars, or suggest symbols like complete quotes and brackets. The outputs of that are usually small. So that can be used via a cmp popup list.

But Copilot and other AI tools are far more to that. They suggest a block of codes, a block of comments, or a long description text. These outputs usually cross multiple lines and construct their own format. If we put these suggestion texts in a single line of a popup list, that will be hard to read and useless.
That's why I think it's far more different than cmp tools.

And Copilot and other similar AI tools are under experiment, for now. how far those AI tools will go, we don't know till we will see. So I think keeping the mature traditional cmp tools to act in their own way, leaving Copilot and other AI tools to act in their new/experiment way, will be a better option, for both edit users and AI tool developers.

**Reply by ogirginc** [reply]
> So I think keeping the mature traditional cmp tools to act in their own way, leaving Copilot and other AI tools to act in their new/experiment way, will be a better option, for both edit users and AI tool developers.

I agree with your thinking. My only "trouble" is copilot & cmp not working with each other out of the box. That's why I wanted to learn what others are doing to make them play nice with each other. :)

**Comment by hrsh7th** [comment]
My prefer way is to not use copilot as completion source.
See this. https://github.com/hrsh7th/nvim-cmp/blob/main/doc/cmp.txt#L812

It is not to same as my actual config but it's similar to that.

**Reply by ogirginc** [reply]
Thank you! I did miss that section on the docs and was using [`zbirenbaum/copilot.lua`](https://github.com/ogirginc/dotfiles/commit/c952fece032a634c93f41da1ac2669dffe6881cc). I will try your way for sure :)

**Reply by hrsh7th** [reply]
This is my actual configuration. Hope to helped.

```lua
      ['<C-o>'] = cmp.mapping(function(fallback)
        local fallback_key = vim.api.nvim_replace_termcodes('<Tab>', true, true, true)
        local resolved_key = vim.fn['copilot#Accept'](fallback)
        if fallback_key == resolved_key then
          cmp.confirm({ select = true })
        else
          vim.api.nvim_feedkeys(resolved_key, 'n', true)
        end
      end),
```

**Reply by ogirginc** [reply]
Thank you 🙌

**Reply by mblarsen** [reply]
Thanks for the snippet. Any clue why this error happen on C-o

```
E5108: Error executing lua: /Users/mbl/.config/nvim/lua/config/plugins/cmp.lua:41: Expected lua string
stack traceback:
        [C]: in function 'nvim_feedkeys'
        /Users/mbl/.config/nvim/lua/config/plugins/cmp.lua:41: in function 'on_keymap'
        /Users/mbl/.local/share/nvim/lazy/nvim-cmp/lua/cmp/core.lua:145: in function 'callback'
        ....local/share/nvim/lazy/nvim-cmp/lua/cmp/utils/keymap.lua:133: in function <....local/share/nvim/lazy/nvim-cmp/lua/cmp/utils/keymap.lua:127>
```
(line 41 is the one with `vim.api.nvim_feedkeys`)

It happens in insert-mode whenever there there are _NO_ suggestions by CoPilot.

**Reply by mellery451** [reply]
same Lua string error for me - did you ever resolve this?

**Reply by mblarsen** [reply]
@mellery451 I did not. Moved away form one-magical-key and have different keys for accepting and moving up and down etc.
